packages/runtime-core/src/apiWatch.ts

```diff
@@ -167,13 +167,15 @@ function doWatch(
 
   let getter: () => any
   let forceTrigger = false
+  let isSourceArray: boolean = false
   if (isRef(source)) {
     getter = () => (source as Ref).value
     forceTrigger = !!(source as Ref)._shallow
   } else if (isReactive(source)) {
     getter = () => source
     deep = true
   } else if (isArray(source)) {
+    isSourceArray = true
```

Comment:
Add `deep = true` in the `if (isReactive(s))` block:
```suggestion
isSourceArray = true
    getter = () =>
      source.map(s => {
        if (isRef(s)) {
          return s.value
        } else if (isReactive(s)) {
          deep = true
          return traverse(s)
        } else if (isFunction(s)) {
          return callWithErrorHandling(s, instance, ErrorCodes.WATCH_GETTER)
        } else {
          __DEV__ && warnInvalidSource(s)
        }
      })
```